Provide plugin for Guava Cache

whl12345 opened the pull request:
link issues #7234

wu-sheng commented:
Hi, you need to follow at least these points
1. Cache plugin should be in the optional list. I noticed `ehcache` plugin in the default plugin, which is a mistake. @mrproliu Look like it was contributed by you, could you move it to an optional plugin with doc updates?
2. Follow the plugin dev doc's last section, we require plugin integration test like all other plugins
3. Documents, `Plugin-list.md`(related to CI process), `Supported-list.md`, and `changes.md` should be updated.

whl12345 commented:
Ok,I will check these points and recommit.

wu-sheng commented:
You need to resolve conflicts first.

whl12345 commented:
> You need to resolve conflicts first.

Ok,I have resolved.

wu-sheng commented:
Please fix these detected CVEs.

wu-sheng commented on apm-sniffer/optional-plugins/guava-cache-plugin/pom.xml:
```diff
@@ -17,28 +17,34 @@
         </dependency>
+
     </dependencies>
+
```
As you support 18-23, could we use 23 in the compiling level. I know this could cause a CVE issue, but it is better to reduce confusion.

wu-sheng commented on docs/en/setup/service-agent/java-agent/Supported-list.md:
```diff
@@ -115,6 +115,7 @@ metrics based on the tracing data.
   * JRE Callable and Runnable (Optional²)
 * Cache
   * [Ehcache](https://www.ehcache.org/) 2.x
+  * [GuavaCache](https://github.com/google/guava) 18.x -> 23.x
```
```suggestion
  * [GuavaCache](https://github.com/google/guava) 18.x -> 23.x (Optional²)
```

wu-sheng commented:
Agent setup doc is not updated, as this is an optional plugin, it should be listed there, and describe why it is optional. Take a reference to how ehcache just did.

whl12345 commented:
Some checks were not successfull, could you tell me what causes these problems?

wu-sheng commented:
Click the details, you could read the logs on your own.

> 2021-07-06T08:21:52.3442672Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (validate) on project apm-guava-cache-plugin: You have 1 Checkstyle violation

dmsolr commented on test/plugin/scenarios/guava-cache-scenario/support-version.list:
```diff
@@ -0,0 +1,17 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+23.0
```
Can this plugin work with the version after 23.0? 
I notice, it is been almost 4 years ago since it was released.

whl12345 replied:
It cannot work with the version after 23.0,because some critical classes were changed.

wu-sheng replied:
As 23 is already a very old version, let's keep limited versions, but true, should not only 23.0

whl12345 replied:
Thanks,I will fix it.

dmsolr commented on docs/en/setup/service-agent/java-agent/Supported-list.md:
```diff
@@ -116,6 +116,7 @@ metrics based on the tracing data.
   * JRE Callable and Runnable (Optional²)
 * Cache
   * [Ehcache](https://www.ehcache.org/) 2.x
+  * [GuavaCache](https://github.com/google/guava) 18.x -> 23.x (Optional²)
```
In commonly, we have to list all versions, we supported, in `support-version.list`

whl12345 replied:
Could we user range to represent multiversion situation. I notice some plugins do not list all versions.

wu-sheng replied:
We pick some typical releases, such as one version per first-level version release.

wu-sheng commented:
> Click the details, you could read the logs on your own.
> 
> > 2021-07-06T08:21:52.3442672Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (validate) on project apm-guava-cache-plugin: You have 1 Checkstyle violation

After you fix this, I will approve to run CIs again.

whl12345 commented:
> > Click the details, you could read the logs on your own.
> > > 2021-07-06T08:21:52.3442672Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (validate) on project apm-guava-cache-plugin: You have 1 Checkstyle violation
> 
> After you fix this, I will approve to run CIs again.

whl12345 commented:
> You have 1 Checkstyle violation

 I have fixed it.Please check it.

wu-sheng commented:
Your plugin test case doesn't pass, please check the logs and verify locally.

whl12345 commented:
> Your plugin test case doesn't pass, please check the logs and verify locally.

It's my problem and I have fixed it.

wu-sheng commented:
> > Your plugin test case doesn't pass, please check the logs and verify locally.
> 
> It's my problem and I have fixed it.

Still failing. You should verify the result locally, we have limited resources on GitHub CIs.

whl12345 commented:
> > > Your plugin test case doesn't pass, please check the logs and verify locally.
> > 
> > 
> > It's my problem and I have fixed it.
> 
> Still failing. You should verify the result locally, we have limited resources on GitHub CIs.

I used skywalking-agent-test-tool to test locally and fixed some promblem. Now its successful.

wu-sheng reviewed (approved):
Generally good. @dmsolr @mrproliu Please recheck.

dmsolr reviewed (approved):
LGTM
